x-pack/plugins/infra/public/components/lens/lens_wrapper.tsx

```diff
@@ -0,0 +1,159 @@
+  loading,
+  query,
+  ...props
+}: LensWrapperProps) => {
+  const [intersectionObserverEntry, setIntersectionObserverEntry] =
+    useState<IntersectionObserverEntry>();
+  const [embeddableLoaded, setEmbeddableLoaded] = useState(false);
+  const [state, setState] = useState({
+    attributes,
+    dateRange,
+    filters,
+    lastReloadRequestTime,
+    query,
+  });
+
+  const ref = useRef<HTMLDivElement>(null);
+  const observerRef = useRef(
+    new IntersectionObserver(([value]) => setIntersectionObserverEntry(value), {
+      root: ref.current,
+    })
+  );
+
+  useEffect(() => {
+    const { current: currentObserver } = observerRef;
+    currentObserver.disconnect();
+    const { current } = ref;
+
+    if (current) {
+      currentObserver.observe(current);
+    }
+
+    return () => currentObserver.disconnect();
+  }, [ref]);
+
+  useEffect(() => {
+    if (intersectionObserverEntry?.isIntersecting) {
+      setState({
+        attributes,
+        dateRange,
+        filters,
+        lastReloadRequestTime,
+        query,
+      });
+    }
+  }, [
+    attributes,
+    dateRange,
+    filters,
+    intersectionObserverEntry?.isIntersecting,
+    lastReloadRequestTime,
+    query,
+  ]);
+
+  const onLoad = useCallback(() => {
+    if (!embeddableLoaded) {
+      setEmbeddableLoaded(true);
+    }
+  }, [embeddableLoaded]);
+
+  const parsedDateRange: TimeRange = useMemo(() => {
+    const { from = state.dateRange.from, to = state.dateRange.to } = parseDateRange(
+      state.dateRange
+    );
+
+    return { from, to };
+  }, [state.dateRange]);
+
+  const isLoading = loading || !state.attributes;
+
+  return (
+    <Container
+      ref={ref}
+      css={css`
+        .echLegend .echLegendList {
+          display: flex;
+        }
+      `}
+    >
+      <>
+        {isLoading && !embeddableLoaded ? (
+          <ChartPlaceholder style={props.style} />
+        ) : (
+          <>
+            {isLoading && <ChartLoadingProgress hasTopMargin={!props.hidePanelTitles} />}
+            <EmbeddableComponentMemo
+              {...props}
+              attributes={state.attributes}
+              filters={state.filters}
+              lastReloadRequestTime={state.lastReloadRequestTime}
+              onLoad={onLoad}
+              query={state.query}
+              timeRange={parsedDateRange}
+              viewMode={ViewMode.VIEW}
+            />
+          </>
+        )}
+      </>
+    </Container>
+  );
+};
+
+const EmbeddableComponentMemo = React.memo(
+  ({
+    attributes,
+    ...props
+  }: Omit<TypedLensByValueInput, 'attributes'> & { attributes: LensAttributes | null }) => {
+    const {
+      services: { lens },
+    } = useKibanaContextForPlugin();
+
+    const EmbeddableComponent = lens.EmbeddableComponent;
+
+    if (!attributes) {
+      return <ChartPlaceholder style={props.style} />;
+    }
+
+    return <EmbeddableComponent {...props} attributes={attributes} />;
+  }
+);
+
+const Container = euiStyled.div`
```

Comment:
Nit: Can we have the styles merged in the styled component or use the emotion `css` also for those styles ( so we have one approach in the component )